### PR TITLE
8303951: Add asserts before record_method_not_compilable where possible

### PR DESCRIPTION
**Pull request by eme64**
I went through all `C2` bailouts, and checked if they are justified to bail out of compilation silently. I added asserts everywhere. Those that were hit, I inspected by hand.

Some of them seem to be justified. There I added comments why they are justified. They are cases that we do not want to handle in `C2`, and that are rare enough so that it probably does not matter.

For the following bailouts I did not add an assert, because it may have revealed a bug:
[JDK-8304328](https://bugs.openjdk.org/browse/JDK-8304328) C2 Bailout "failed spill-split-recycle sanity check" reveals hidden issue with RA

Note:
[JDK-8303466](https://bugs.openjdk.org/browse/JDK-8303466) C2: COMPILE SKIPPED: malformed control flow - only one IfProj
That bug bug was the reason for this RFE here. I added the assert for "malformed control flow". After this RFE here, that Bug will run into the assert on debug builds.

I ran `tier1-6` and stress testing. Now running `tier7-9`.

Filed a follow-up RFE to do the same for `BAILOUT` in `C1`: [JDK-8304532](https://bugs.openjdk.org/browse/JDK-8304532).

Note: the philosophy here is rather to have an assert too much that triggers in the future. Then we can re-evaluate and weaken or remove it - or maybe we find a bug that we can fix.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303951](https://bugs.openjdk.org/browse/JDK-8303951): Add asserts before record_method_not_compilable where possible

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [28d41ffe](https://git.openjdk.org/jdk/pull/13038/files/28d41ffef70d5b040c0126b2b7e9097f330c2ff2)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13038/head:pull/13038` \
`$ git checkout pull/13038`

Update a local copy of the PR: \
`$ git checkout pull/13038` \
`$ git pull https://git.openjdk.org/jdk.git pull/13038/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13038`

View PR using the GUI difftool: \
`$ git pr show -t 13038`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13038.diff">https://git.openjdk.org/jdk/pull/13038.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
General suggestion is to print more info on bailouts to help debugging.

File: src/hotspot/share/opto/compile.cpp [hunk 1]
```diff
@@ -753,7 +753,11 @@ Compile::Compile( ciEnv* ci_env, ciMethod* target, int osr_bci,
     }
     if (failing())  return;
     if (cg == nullptr) {
-      record_method_not_compilable("cannot parse method");
+      const char* reason = InlineTree::check_can_parse(method());
+      assert(reason != nullptr, "cannot parse method: why?");
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/compile.cpp:757]: `reason` is `nullptr` when the asser fails. I now say `expect reason for parse failure`.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Review by TobiHartmann — COMMENTED** [review]
Looks good overall, some comments below.

File: src/hotspot/share/opto/parse1.cpp [hunk 2]
```diff
@@ -208,11 +208,14 @@ void Parse::load_interpreter_state(Node* osr_buf) {
...
   // of loops in catch blocks or loops which branch with a non-empty stack.
   if (sp() != 0) {
+    // Bailout. But we should probably kick into normal compilation?
     C->record_method_not_compilable("OSR starts with non-empty stack");
     return;
   }
   // Do not OSR inside finally clauses:
   if (osr_block->has_trap_at(osr_block->start())) {
+    assert(false, "OSR starts with an immediate trap");
+    // Bailout. But we should probably kick into normal compilation?
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/parse1.cpp:218]: It sounds likely, but I never saw it happen, up to tier9. So I suggest we leave the assert in, and once it is triggered, we can decide to remove it again. But then we also have an example, and we can check it in as a test, to justify the bailout.

**Review by TobiHartmann — APPROVED** [review]
Thanks for making these changes, looks good to me.

**Comment by eme64** [comment]
@vnkozlov @TobiHartmann thanks for the help and reviews!
/integrate

**Pull request closed without merging**